File: src/jvmTest/kotlin/community/flock/kotlinx/rgxgen/CombinedCaseInsensitiveTests.java
package community.flock.kotlinx.rgxgen;

import community.flock.kotlinx.rgxgen.config.RgxGenOption;
import community.flock.kotlinx.rgxgen.config.RgxGenProperties;
import community.flock.kotlinx.rgxgen.data.TestPatternCaseInsensitive;
import community.flock.kotlinx.rgxgen.testutil.TestingUtilities;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

public class CombinedCaseInsensitiveTests extends CombinedTestTemplate<TestPatternCaseInsensitive> {
    public static Stream<TestPatternCaseInsensitive> getAllPatterns() {
        return Arrays.stream(TestPatternCaseInsensitive.values());
    }

    @ParameterizedTest
    @MethodSource("getAllPatterns")
    public void countTest(TestPatternCaseInsensitive testPattern) {
        assumeTrue(testPattern.hasEstimatedCount());
        RgxGenProperties properties = new RgxGenProperties();
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true);
        RgxGen rgxGen = RgxGen.parse(properties, testPattern.getPattern());
        assertEquals(testPattern.getEstimatedCount(), rgxGen.getUniqueEstimation());
    }

    @ParameterizedTest
    @MethodSource("getAllPatterns")
    public void generateUniqueTest(TestPatternCaseInsensitive testPattern) {
        assumeTrue(testPattern.hasAllUniqueValues());
        RgxGenProperties properties = new RgxGenProperties();
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true);
        RgxGen rgxGen = RgxGen.parse(properties, testPattern.getPattern());
        assertEquals(testPattern.getAllUniqueValues(), TestingUtilities.iteratorToList(rgxGen.iterateUnique()));
    }

    @ParameterizedTest
    @MethodSource("getAllPatterns")
    public void classRgxGenCaseInsensitiveTest(TestPatternCaseInsensitive testPattern) {
        RgxGenProperties properties = new RgxGenProperties();
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true);
        RgxGen rgxGen = RgxGen.parse(properties, testPattern.getPattern());
        Iterator<String> iterator = rgxGen.stream();
        Iterable<String> iterable = () -> iterator;
        List<String> strings = StreamSupport.stream(iterable.spliterator(), false)
                .limit(1000)
                .collect(Collectors.toList());
        Pattern caseSensitivePattern = Pattern.compile(testPattern.getPattern());
        boolean atLeastOneCaseSensitiveMismatch = false;
        for (String generated : strings) {
            boolean result = isValidGenerated(testPattern, generated, Pattern.CASE_INSENSITIVE);
            boolean caseSensitiveMatches = !caseSensitivePattern.matcher(generated).matches();
            assertTrue(result, "Text: '" + generated + "' does not match pattern " + testPattern.getPattern());
            atLeastOneCaseSensitiveMismatch = atLeastOneCaseSensitiveMismatch || caseSensitiveMatches;
        }
        assertTrue(atLeastOneCaseSensitiveMismatch);
    }
}

File: src/jvmTest/kotlin/community/flock/kotlinx/rgxgen/CombinedCaseInsensitiveTests.kt
package community.flock.kotlinx.rgxgen

import community.flock.kotlinx.rgxgen.RgxGen.Companion.parse
import community.flock.kotlinx.rgxgen.config.RgxGenOption
import community.flock.kotlinx.rgxgen.config.RgxGenProperties
import community.flock.kotlinx.rgxgen.data.TestPatternCaseInsensitive
import community.flock.kotlinx.rgxgen.testutil.TestingUtilities.iteratorToList
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.util.Arrays
import java.util.regex.Pattern
import java.util.stream.Collectors
import java.util.stream.Stream
import java.util.stream.StreamSupport

class CombinedCaseInsensitiveTests : CombinedTestTemplate<TestPatternCaseInsensitive?>() {
    @ParameterizedTest
    @MethodSource("getAllPatterns")
    fun countTest(testPattern: TestPatternCaseInsensitive) {
        Assumptions.assumeTrue(testPattern.hasEstimatedCount())
        val properties: RgxGenProperties<Any> = RgxGenProperties<Any>()
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true)
        val rgxGen = parse(properties, testPattern.getPattern())
        Assertions.assertEquals(testPattern.getEstimatedCount(), rgxGen.uniqueEstimation)
    }

    @ParameterizedTest
    @MethodSource("getAllPatterns")
    fun generateUniqueTest(testPattern: TestPatternCaseInsensitive) {
        Assumptions.assumeTrue(testPattern.hasAllUniqueValues())
        val properties: RgxGenProperties<Any> = RgxGenProperties<Any>()
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true)
        val rgxGen = parse(properties, testPattern.getPattern())
        Assertions.assertEquals(testPattern.getAllUniqueValues(), iteratorToList<String?>(rgxGen.iterateUnique()!!))
    }

    @ParameterizedTest
    @MethodSource("getAllPatterns")
    fun classRgxGenCaseInsensitiveTest(testPattern: TestPatternCaseInsensitive) {
        val properties: RgxGenProperties<Any> = RgxGenProperties<Any>()
        RgxGenOption.CASE_INSENSITIVE.setInProperties(properties, true)
        val rgxGen = parse(properties, testPattern.getPattern())
        val iterator = rgxGen.stream()
        val iterable = Iterable { iterator }
        val strings = StreamSupport.stream(iterable.spliterator(), false)
            .limit(1000)
            .collect(Collectors.toList())
        val caseSensitivePattern = Pattern.compile(testPattern.getPattern())
        var atLeastOneCaseSensitiveMismatch = false
        for (generated in strings) {
            val result = isValidGenerated(testPattern, generated, Pattern.CASE_INSENSITIVE)
            val caseSensitiveMatches = !caseSensitivePattern.matcher(generated).matches()
            Assertions.assertTrue(
                result,
                "Text: '" + generated + "' does not match pattern " + testPattern.getPattern()
            )
            atLeastOneCaseSensitiveMismatch = atLeastOneCaseSensitiveMismatch || caseSensitiveMatches
        }
        Assertions.assertTrue(atLeastOneCaseSensitiveMismatch)
    }

    companion object {
        @JvmStatic
        val allPatterns: Stream<TestPatternCaseInsensitive>
            get() = Arrays.stream(TestPatternCaseInsensitive.entries.toTypedArray())
    }
}
